src/v/raft/state_machine_manager.cc

```diff
@@ -362,7 +362,17 @@ ss::future<> state_machine_manager::try_apply_in_foreground() {
         auto last_applied = co_await std::move(reader).consume(
           batch_applicator(default_ctx, machines, _as, _log),
           model::no_timeout);
-
+        if (model::next_offset(last_applied) == _next) {
+            vlog(
+              _log.warn,
+              "no progress has been made during state machine apply. Current "
+              "next offset: {}",
+              _next);
+            /**
+             * If no progress has been made, yield to prevent busy looping
+             */
+            co_await ss::sleep_abortable(100ms, _as);
```

Comment:
this makes sense but we should probably have some mechanism to detect stm stalls I think, via some lag metrics. In this case we are lucky this manifested as high reactor utilization so we had to dig deep.

Reply:
you are right, i was thinking about that, so far i added a `WARN` log message, metric seems like a great idea. Maybe a follow up PR ?

Reply:
ofcourse